ksql-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java

```diff
@@ -286,10 +286,11 @@ private PlanNode buildSourceNode() {
         new PlanNodeId("Join"),
         analysis.getSelectExpressions(),
         joinInfo.get().getType(),
-        leftSourceNode,
-        rightSourceNode,
-        joinInfo.get().getLeftJoinField(),
-        joinInfo.get().getRightJoinField(),
+        // it is always safe to build the repartition node - this operation will be
+        // a no-op if a repartition is not required. if the source is a table, and
+        // a repartition is needed, then an exception will be thrown
+        buildRepartitionNode(leftSourceNode, joinInfo.get().getLeftJoinExpression()),
+        buildRepartitionNode(rightSourceNode, joinInfo.get().getRightJoinExpression()),
```

Comment:
> IMO the transformation performed by the nodes in the logical plan should correspond closely to the semantics of the corresponding transformation in the query

My knowledge on SQL theory and relational algebra is pretty minimal, but I feel that the fact that the nodes mirror the SQL statements is coincidental (a reflection of how new our codebase is) not characteristic. Most databases have a step that optimizes the logical plan independent of the physical plan so the output of the logical planner may only loosely resemble the original SQL query. Otherwise, why have the extra step to translate between the query nodes and the logical plan nodes?

Instead, I think the logical nodes should define steps that describe the logical transformation and they should constitute the smallest reusable unit of logical work - otherwise reusing (composing) them and optimizing complex queries becomes unmanageable. The `LogicalPlanner`'s job is to turn a SQL query into a logical plan.

> The semantics of the repartition for the join might be different from the semantics of the repartition needed to perform a `PARTITION BY` ... If `PARTITION BY` semantics support adding columns in the `PARTITION BY`, those columns should get some generated name. But it wouldn't make sense to add those columns for the repartition done here.

This falls pretty well into my view of logical plan. In my view, if `PARTITION BY` supports adding columns then a single `PARTITION BY` should map to two logical nodes - a projection node (or a rename) and a partition node. 

> The repartition here is really an implementation detail of the join.

I think things get a little tricky with `PARTITION BY` because it's the only step that spans the logical and the physical world. Really, we don't have a "RekeyNode" (which is perhaps a better name for the `RepartitionNode`) - but we _have_ defined our Joins as only working on streams with the same key (and this a logical, not a physical, requirement).

> If we did that it would be awkward to have the original rekey in these repartition nodes and the post-join rekey in JoinNode

It wouldn't be in the JoinNode - it would be yet another `RepartitionNode` (which, again, should probably be named `RekeyNode`).